src/sparse_blas/backends/mklcpu/mklcpu_basic.cpp

```diff
@@ -0,0 +1,93 @@
+/*******************************************************************************
+* Copyright 2023 Codeplay Software Ltd.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions
+* and limitations under the License.
+*
+*
+* SPDX-License-Identifier: Apache-2.0
+*******************************************************************************/
+
+// MKLCPU backend includes
+// These includes define their own oneapi::mkl::sparse namespace with some of the types that are used here: matrix_handle_t, index_base.
+#include <oneapi/mkl/spblas.hpp>
+
+#include "oneapi/mkl/sparse_blas/detail/helper_types.hpp"
+#include "oneapi/mkl/sparse_blas/detail/mklcpu/onemkl_sparse_blas_mklcpu.hpp"
+
+// Include are set up so that oneapi::mkl::sparse namespace refers to the MKLCPU namespace in this file.
+// oneapi::mkl::sparse::detail namespace refers to the oneMKL interface namespace.
+
+namespace oneapi::mkl::sparse::detail {
+
+auto get_handle(detail::matrix_handle **handle) {
+    return reinterpret_cast<matrix_handle_t *>(handle);
+}
```

Comment:
So `oneapi::mkl::sparse::detail::matrix_handle **handle` is being `reinterpret_cast<>`ed to `oneapi::mkl::sparse::matrix_handle_t *` _without_ the `detail` namespace here? Does that mean that `oneapi::mkl::sparse::matrix_handle_t` is the public-facing `matrix_handle_t` of the oneMKL Interfaces repo, and `oneapi::mkl::sparse::detail::matrix_handle_t` is the handle that is coming in from the Intel oneMKL product?

Reply:
Not quite:

- `oneapi::mkl::sparse::detail::matrix_handle_t` does not exist. oneMKL interface only defines the type `oneapi::mkl::sparse::detail::matrix_handle` and `oneapi::mkl::sparse::matrix_handle_t` as an alias to `* detail::matrix_handle`. The latter is the public-facing type as you said.
- In this file however things are different. The type `matrix_handle_t` is defined by the close-source header from oneMKL product and only `detail::matrix_handle` from oneMKL interface is visible.
- This function is in the `oneapi::mkl::sparse::detail` namespace but it can use the types from the parent namespaces. In this case `matrix_handle_t` refers to `oneapi::mkl::sparse::matrix_handle_t`. I've added the namespace explicitly in https://github.com/oneapi-src/oneMKL/pull/374/commits/97d3bbd so hopefully that's clearer.